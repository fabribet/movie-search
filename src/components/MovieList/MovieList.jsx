import React from 'react'
import PropTypes from 'prop-types'
import styles from './styles.module.scss'

import MovieBox from '../MovieBox'

/**
 * MovieList - React component.
 * Renders a lists of movies.
 *
 * Properties
 * - imgUrl {String} - The base url to get movie posters. {Required}
 * - movie {Object} - The movie information
 *    - title
 *    - overview
 *    - vote_average
 *    - poster_path
 */
function MovieList(props) {
  return (
    <div className={styles.movieList}>
      <div className={styles.wrapper}>
        <h2 className={styles.title}>{props.title}</h2>
        <div className={styles.moviesContainer}>
          {props.movies && props.movies.length
            ? props.movies.map((movie, index) => <MovieBox key={index} movie={movie} favorite={!!(props.favorites[movie.id])}/>)
            : 'There are no movies that match the criteria, try again.'
          }
        </div>
      </div>
    </div>
  )
}

MovieList.propTypes = {
    favorites: PropTypes.object,
    title: PropTypes.string,
    movies: PropTypes.arrayOf(
      PropTypes.shape({
        title: PropTypes.string.isRequired,
        vote_average: PropTypes.number,
        overview: PropTypes.string,
        poster_path: PropTypes.string.isRequired
      }).isRequired
    ).isRequired
}

export default MovieList