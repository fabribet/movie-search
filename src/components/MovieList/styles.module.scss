@import '../../scss/colors.scss';

.movieList {
  padding: 10px;
  text-align: center;
  display: flex;
  flex-direction: column;

  .wrapper {
    width: 90%;
    max-width: 1000px;
    padding: 0 10px 10px 10px;
    align-self: center;

    .title {
      margin-left: 15px;
      text-align: left;
    }

    .moviesContainer {
      margin-top: 10px;
      display: flex;
      flex-wrap: wrap;
      justify-content: flex-start;
    }
  }
}

