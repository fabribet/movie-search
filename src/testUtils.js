import configureStore from 'redux-mock-store'

// Mock the canvas usage for external library usage on tests.
export function mockCanvas () {
  const createElement = document.createElement.bind(document)
  document.createElement = tagName => {
    const element = createElement(tagName)
    if (tagName === 'canvas') {
      element.getContext = () => ({})
    }
    return element
  }
}

/**
 * Create Mock store for testing
 */
const mockStore = configureStore([])
export function createMockStore (initialState = {
  Favorites: {
    favorites: {}
  },
  TMDB: {
    apiConfig: {
      loading: false,
      error: null,
      data: {
        images: {
          base_url: ''
        }
      }
    }
  }
}) {
  return mockStore(initialState)
}

export const FAVORITES_DATA = {
  1: {
    id: 1,
    title: 'Movie 1',
    poster_path: '/asdasdas.jpg'
  },

  2: {
    id: 2,
    title: 'Movie 2',
    poster_path: '/asdasdas2.jpg'
  }
}
