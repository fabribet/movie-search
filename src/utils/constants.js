
/**
 * Text displayed when no favorites have been added
 */
export const EMPTY_FAVORITES_TXT = 'Add favorites to your list to see them displayed here.'

/**
 * Text displayed when no movies have been found
 */
export const EMPTY_MOVIE_LIST_TXT = 'There are no movies that match the criteria, try again.'
