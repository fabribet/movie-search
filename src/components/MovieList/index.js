import { connect } from 'react-redux'
import MovieList from './MovieList'

const mapStateToProps = state => ({
  favorites: state.Favorites.favorites
})

export default connect(mapStateToProps)(MovieList)
