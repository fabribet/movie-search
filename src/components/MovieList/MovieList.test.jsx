import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'

import MovieList from './MovieList'
import { mockCanvas, createMockStore, FAVORITES_DATA } from './../../testUtils'
import { EMPTY_MOVIE_LIST_TXT } from '../../utils/constants'

mockCanvas()
const store = createMockStore()
const TITLE = 'PAGE_TITLE'
const MOVIES = [
  {
    id: FAVORITES_DATA[1].id,
    title: FAVORITES_DATA[1].title,
    vote_average: 5,
    vote_count: 83,
    overview: 'An overview',
    poster_path: FAVORITES_DATA[1].poster_path
  },
  {
    id: 3,
    title: 'Third Title',
    vote_average: 4,
    vote_count: 811,
    overview: 'An overview of the third',
    poster_path: '/mythirdpath.jpg'
  },
  {
    id: 4,
    title: 'Fourth Title',
    vote_average: 7,
    vote_count: 111,
    overview: 'An overview of the fourth',
    poster_path: '/myfourthpath.jpg'
  }
]

const movieList =
  <Provider store={store}>
    <MovieList favorites={FAVORITES_DATA} title={TITLE} movies={MOVIES} />
  </Provider>

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(
    movieList,
    div
  )
})

it('Renders MovieList correctly', () => {
  const tree = renderer
    .create(movieList)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Renders a "no movies found" message when nothing was found', () => {
  const { queryByText } = render(
    <Provider store={store}>
      <MovieList favorites={{}} title={TITLE} movies={[]} />
    </Provider>
  )

  expect(queryByText(EMPTY_MOVIE_LIST_TXT)).not.toBeNull()
})

it('Renders the list and no text is displayed', () => {
  const { queryByText } = render(
    <Provider store={store}>
      <MovieList favorites={FAVORITES_DATA} title={TITLE} movies={MOVIES} />
    </Provider>
  )

  expect(queryByText(EMPTY_MOVIE_LIST_TXT)).toBeNull()
})
