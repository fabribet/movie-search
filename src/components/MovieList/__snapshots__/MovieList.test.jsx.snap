// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders MovieList correctly 1`] = `
<div
  className="movieList"
>
  <div
    className="wrapper"
  >
    <h2
      className="title"
    >
      PAGE_TITLE
    </h2>
    <div
      className="moviesContainer"
    >
      <div
        className="box"
        onClick={[Function]}
      >
        <div
          className="img"
          style={
            Object {
              "backgroundImage": "url(movie-poster.png)",
            }
          }
        />
        <div
          className="movieInfo"
        >
          <div
            className="titleWrapper"
          >
            <h4
              className="title"
            >
              <span
                width={0}
              >
                <span />
                <span>
                  Movie 1
                </span>
                <span
                  style={
                    Object {
                      "left": 0,
                      "position": "fixed",
                      "top": 0,
                      "visibility": "hidden",
                    }
                  }
                >
                  ...
                </span>
              </span>
            </h4>
            <div
              className="favorite solid"
              id="favoriteBtn"
              onClick={[Function]}
              title="Remove favorite"
            />
          </div>
          <div
            className="ratingWrapper"
          >
            Rating:
            <span>
              <span
                className="rating normalRating"
              >
                5
              </span>
              Votes:
              <span
                className="voteCount"
              >
                83
              </span>
            </span>
          </div>
          <div
            className="overview"
          >
            <span
              width={0}
            >
              <span />
              <span>
                An overview
              </span>
              <span
                style={
                  Object {
                    "left": 0,
                    "position": "fixed",
                    "top": 0,
                    "visibility": "hidden",
                  }
                }
              >
                ...
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        className="box"
        onClick={[Function]}
      >
        <div
          className="img"
          style={
            Object {
              "backgroundImage": "url(movie-poster.png)",
            }
          }
        />
        <div
          className="movieInfo"
        >
          <div
            className="titleWrapper"
          >
            <h4
              className="title"
            >
              <span
                width={0}
              >
                <span />
                <span>
                  Third Title
                </span>
                <span
                  style={
                    Object {
                      "left": 0,
                      "position": "fixed",
                      "top": 0,
                      "visibility": "hidden",
                    }
                  }
                >
                  ...
                </span>
              </span>
            </h4>
            <div
              className="favorite empty"
              id="favoriteBtn"
              onClick={[Function]}
              title="Add favorite"
            />
          </div>
          <div
            className="ratingWrapper"
          >
            Rating:
            <span>
              <span
                className="rating lowRating"
              >
                4
              </span>
              Votes:
              <span
                className="voteCount"
              >
                811
              </span>
            </span>
          </div>
          <div
            className="overview"
          >
            <span
              width={0}
            >
              <span />
              <span>
                An overview of the third
              </span>
              <span
                style={
                  Object {
                    "left": 0,
                    "position": "fixed",
                    "top": 0,
                    "visibility": "hidden",
                  }
                }
              >
                ...
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        className="box"
        onClick={[Function]}
      >
        <div
          className="img"
          style={
            Object {
              "backgroundImage": "url(movie-poster.png)",
            }
          }
        />
        <div
          className="movieInfo"
        >
          <div
            className="titleWrapper"
          >
            <h4
              className="title"
            >
              <span
                width={0}
              >
                <span />
                <span>
                  Fourth Title
                </span>
                <span
                  style={
                    Object {
                      "left": 0,
                      "position": "fixed",
                      "top": 0,
                      "visibility": "hidden",
                    }
                  }
                >
                  ...
                </span>
              </span>
            </h4>
            <div
              className="favorite empty"
              id="favoriteBtn"
              onClick={[Function]}
              title="Add favorite"
            />
          </div>
          <div
            className="ratingWrapper"
          >
            Rating:
            <span>
              <span
                className="rating goodRating"
              >
                7
              </span>
              Votes:
              <span
                className="voteCount"
              >
                111
              </span>
            </span>
          </div>
          <div
            className="overview"
          >
            <span
              width={0}
            >
              <span />
              <span>
                An overview of the fourth
              </span>
              <span
                style={
                  Object {
                    "left": 0,
                    "position": "fixed",
                    "top": 0,
                    "visibility": "hidden",
                  }
                }
              >
                ...
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
